nexus/db-queries/src/db/datastore/address_lot.rs

```diff
@@ -0,0 +1,333 @@
+use crate::db::error::public_error_from_diesel_pool;
+use crate::db::error::ErrorHandler;
+use crate::db::error::TransactionError;
+use crate::db::model::Name;
+use crate::db::model::{AddressLot, AddressLotBlock, AddressLotRsvdBlock};
+use crate::db::pagination::paginated;
+use async_bb8_diesel::{
+    AsyncConnection, AsyncRunQueryDsl, Connection, ConnectionError, PoolError,
+};
+use chrono::Utc;
+use diesel::result::Error as DieselError;
+use diesel::{ExpressionMethods, QueryDsl, SelectableHelper};
+use diesel_dtrace::DTraceConnection;
+use ipnetwork::IpNetwork;
+use nexus_types::external_api::params;
+use nexus_types::identity::Resource;
+use omicron_common::api::external::http_pagination::PaginatedBy;
+use omicron_common::api::external::{
+    CreateResult, DataPageParams, DeleteResult, Error, ListResultVec,
+    LookupType, NameOrId, ResourceType,
+};
+use ref_cast::RefCast;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct AddressLotCreateResult {
+    pub lot: AddressLot,
+    pub blocks: Vec<AddressLotBlock>,
+}
+
+impl DataStore {
+    pub async fn address_lot_create(
+        &self,
+        opctx: &OpContext,
+        params: &params::AddressLotCreate,
+    ) -> CreateResult<AddressLotCreateResult> {
+        use db::schema::address_lot::dsl as lot_dsl;
+        use db::schema::address_lot_block::dsl as block_dsl;
+
+        self.pool_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
+                let lot = AddressLot::new(&params.identity, params.kind.into());
+
+                let db_lot: AddressLot =
+                    diesel::insert_into(lot_dsl::address_lot)
+                        .values(lot)
+                        .returning(AddressLot::as_returning())
+                        .get_result_async(&conn)
+                        .await?;
+
+                let blocks: Vec<AddressLotBlock> = params
+                    .blocks
+                    .iter()
+                    .map(|b| {
+                        AddressLotBlock::new(
+                            db_lot.id(),
+                            b.first_address.into(),
+                            b.last_address.into(),
+                        )
+                    })
+                    .collect();
+
+                let mut db_blocks: Vec<AddressLotBlock> = Vec::new();
+                for b in &blocks {
+                    let db_block =
+                        diesel::insert_into(block_dsl::address_lot_block)
+                            .values(b.clone())
+                            .returning(AddressLotBlock::as_returning())
+                            .get_result_async(&conn)
+                            .await?;
+                    db_blocks.push(db_block);
+                }
+
+                Ok(AddressLotCreateResult { lot: db_lot, blocks: db_blocks })
+            })
+            .await
+            .map_err(|e| match e {
```

Comment:
What does the user see if the transaction fails because of a conflict? I mean, what if we try to concurrently create two lots that share a block?